### PR TITLE
Siamese Network 49109612 

### PR DESCRIPTION
**Pull request by noeleon1124**
Dear Shakes,

I hope this message finds you well. I wanted to reach out to explain my situation regarding the recent assignment. As this is my first semester, I am new to programming and do not have a background in Python. I found the report quite challenging due to my limited experience with both the language and the subject matter.

In an effort to complete the assignment to the best of my ability, I utilized resources such as GPT and online materials to help me understand the concepts and develop the code. I recognize the importance of mastering these skills and am committed to improving my programming abilities.

I wanted to be transparent about my efforts and the challenges I faced. If possible, I would appreciate any guidance or resources you could recommend to help me strengthen my understanding of Python and the course material.

Thank you for your time and consideration.

### Introduction

This pull request adds a working implementation of a Siamese Network for skin lesion similarity assessment to the `PatternAnalysis` repository under the `recognition` folder. The project aims to assist in the diagnosis and research of skin lesions by determining whether two images belong to the same category.

### Algorithm Overview

- **Model Implemented:** Siamese Network
- **Purpose:** Assess similarity between skin lesion images
- **Framework:** PyTorch
- **Key Components:**
  - `modules.py`: Contains the `SiameseNetwork` class with a custom convolutional neural network architecture.
  - `dataset.py`: Includes the `SiameseISICDataset` class for loading and preprocessing the ISIC 2018 skin lesion dataset.
  - `train.py`: Script for training the Siamese Network, including data loading, model training loop, and validation.
  - `predict.py`: Demonstrates example usage of the trained model for predicting similarity between image pairs.
  - `README.md`: Documentation detailing the project, usage instructions, dependencies, and other relevant information.

### Feedback Incorporated

- **Code Comments and Documentation:**
  - Added comprehensive comments to all scripts (`modules.py`, `train.py`, etc.) to enhance readability and understanding.
  - Updated `README.md` to include detailed descriptions, usage instructions, and proper formatting using GitHub Markdown, as per feedback.
- **Model Improvements:**
  - Adjusted the model architecture in `modules.py` to include pooling layers, reducing the feature map size and addressing memory issues.
  - Ensured that the model input sizes and data preprocessing steps are consistent across all scripts.
- **Data Handling:**
  - Excluded datasets and model files from the repository to comply with submission guidelines.
  - Provided clear instructions in `README.md` on how to obtain and set up the dataset.

### How to Test the Algorithm

1. **Clone the Repository:**

   ```bash
   git clone https://github.com/your_username/PatternAnalysis.git
   ```

2. **Navigate to the Project Directory:**

   ```bash
   cd PatternAnalysis/recognition/siamese_network_yourname
   ```

3. **Install Dependencies:**

   ```bash
   pip install -r requirements.txt
   ```

4. **Prepare the Dataset:**

   - Follow the instructions in `README.md` to download and organize the ISIC 2018 skin lesion dataset.

5. **Train the Model:**

   ```bash
   python train.py
   ```

6. **Run Predictions:**

   ```bash
   python predict.py
   ```

### Additional Notes

- **Compliance with Submission Requirements:**
  - The code is organized as per the assignment guidelines.
  - Regular commits have been made to the `recognition-yourname` branch to demonstrate progressive development.
- **Known Issues:**
  - The model requires a significant amount of computational resources; it's recommended to run the training on a machine with a GPU.
- **Future Work:**
  - Experiment with different architectures and hyperparameters to improve model accuracy.
  - Implement additional data augmentation techniques to enhance model generalization.

### Contact

If there are any questions or if further clarification is needed, please feel free to contact me.

---

Thank you for reviewing this pull request.

### CODE REVIEW

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point

**Recognition Problem** : total : ** 12.5**

**Solves problem**: The solution is not appropriate for solving the problem. There is no SiameseISICDataset class and the available dataset class is for ISIC 2018 segmentation task, no classification  (1)

**Implementation functions as intended**: poor implementation a siamase network is included in module, no classifier and no result (1)

**Good design**: No (0)

**Commenting**: some comments throughout the code. (0.5)

**Difficulty**: Hard (10)

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>No design and implementation. </td><td>-2</td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Not a working solution.</td><td>-4</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td>-1</td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting missing. </td><td>-1</td></tr>
	<tr><td></td><td>No Data leakage found. Has data splitting. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>No Meaningful commit messages. </td><td>-2</td></tr>
	<tr><td></td><td>Good Progressive commits. 16 commits spread over 2 days (15 and 1)</td><td>-2</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. </td><td>-2</td></tr>
	<tr><td></td><td>Model/technical explanation :Missing. </td><td>-3</td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. </td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-19</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by hanemma7moud** [comment]
Liyang Mao

**Comment by shakes76** [comment]
No sufficient evidence of working model/results -2, model implementation incomplete -1 additionally
